perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
 
-    :param list func: the list of function names that will be traced
-    :param list func_sampled: the list of function names with specified sampling
-    :param list static: the list of static probes that will be traced
-    :param list static_sampled: the list of static probes with specified sampling
-    :param kwargs: additional configuration parameters
+    :return tuple: updated global_sampling, extracted functions or {}, extracted static probes or {}
+    """
+    # Set the default sampling if sampled strategy is required but sampling is not provided
+    if method in ('u_sampled', 'a_sampled') and global_sampling == 1:
+        global_sampling = _DEFAULT_SAMPLE
+
+    # Extract the functions and static probes if needed
+    func = _extract_functions(binary, method, global_sampling) if method != 'custom' else {}
+    static = _extract_static(binary, global_sampling) if method != 'custom' and with_static else {}
+
+    return global_sampling, func, static
 
-    :returns kwargs: the updated dictionary with post processed rules
+
+def _merge_probes(specified, specified_sampled, extracted, global_sampling):
+    """Merges the probe lists without / with specified sampling and extracted functions into one
+    list with unified sampling specification
+
+    :param list specified: list of strings that represent probe names
+    :param list specified_sampled: list of tuples that contain 0) probe name, 1) probe sampling
+    :param dict extracted: dictionary of extracted probe specification in unified format
+    :param int global_sampling: the global sampling value that is applied to all probes
+                                without specified sampling
+    :return dict: dict of probes in unified format (dictionaries with 'name' and 'sample' keys)
     """
-    # Apply global sampling for sampled strategies
-    if ((kwargs['method'] == 'u_sampled' or kwargs['method'] == 'a_sampled')
-            and kwargs['global_sampling'] == 1):
-        kwargs['global_sampling'] = _DEFAULT_SAMPLE
-
-    # Create probe locations: process separately user and extracted locations, then merge them
-    kwargs['func'] = _merge_extracted_with_custom(
-        _remove_duplicate_probes(_extract_functions(**kwargs)),
-        _remove_duplicate_probes(_merge_probes_lists(func, func_sampled,
-                                                     kwargs['global_sampling'])))
-    kwargs['static'] = []
-    if kwargs['with_static']:
-        kwargs['static'] = _merge_extracted_with_custom(
-            _remove_duplicate_probes(_pair_rules(_extract_static_probes(**kwargs))),
-            _remove_duplicate_probes(_pair_rules(_merge_probes_lists(static, static_sampled,
-                                                                     kwargs['global_sampling']))))
-    return kwargs
+    # Add global sampling (default 0) to the probes without sampling specification
+    probes = {probe: {'name': probe, 'sample': global_sampling} for probe in specified}
```

Comment:
Just nitpicking suggestion: maybe if you expand it to more lines, it would look less cramped. But no need to fix this.
```suggestion
    probes = {
        probe: {
            'name': probe, 
            'sample': global_sampling
        } for probe in specified
    }
```